### ISSUE TITLE
Help understanding copy_generator.py

### ISSUE THREAD

**Issue by Skylixia** [OpenNMT/OpenNMT-py#2050]
Hi,

I'm trying to understand the implementation of the copy mechanism but have a few questions. 
```
def forward(self, hidden, attn, src_map):       
        batch_by_tlen, _ = hidden.size()
        batch_by_tlen_, slen = attn.size()
        slen_, batch, cvocab = src_map.size()
        aeq(batch_by_tlen, batch_by_tlen_)
        aeq(slen, slen_)

        # Original probabilities.
        logits = self.linear(hidden)
        logits[:, self.pad_idx] = -float('inf')
        prob = torch.softmax(logits, 1)

        # Probability of copying p(z=1) batch.
        p_copy = torch.sigmoid(self.linear_copy(hidden))
        # Probability of not copying: p_{word}(w) * (1 - p(z))
        out_prob = torch.mul(prob, 1 - p_copy)
        mul_attn = torch.mul(attn, p_copy)
        copy_prob = torch.bmm(
            mul_attn.view(-1, batch, slen).transpose(0, 1),
            src_map.transpose(0, 1)
        ).transpose(0, 1)
        copy_prob = copy_prob.contiguous().view(-1, cvocab)
        return torch.cat([out_prob, copy_prob], 1)
```
1. 
From the original [paper](https://nlp.stanford.edu/pubs/see2017get.pdf), the generation probability p_gen ∈ [0,1] for timestep t is calculated from the context vector h∗t, the decoder state st and the decoder input xt:
p_gen = σ(wT_h∗ h∗t +wT_s st +wT_x xt +b_ptr)
This is used as a switch to chose whether to generate or copy from the input. This probability is then multiplied to the vocabulary distribution and 1-p_gen to the attention distribution (probability of a word from the input).

In the code, I can see the vocab distribution:  prob = torch.softmax(logits, 1) and the attention distribution: p_copy = torch.sigmoid(self.linear_copy(hidden)). 
But I can't figure out how the switch p_gen is implemented.

2.
src_map:  A sparse indicator matrix mapping each source word to
               its index in the "extended" vocab containing.
How is this mapping built ? Where is it implemented in the code ?

**Comment by h324yang** [comment]
I think the implementation is not the same, and I also want to confirm this with the authors. 
It seems that p_copy is used as the switch here, i.e., ``p_gen''. 
If this is the case, it means it's only conditional on decoder state, i.e., ``hidden'', which is different from the paper. 

A follow-up question is that I can't find where out_prob  and copy_prob are summed up. 
Instead, they are concatenated.
Mathematically, here seems NOT equivalent to Pointer Generator Networks, either?

**Comment by vince62s** [comment]
indeed, the implementation is not the same as the paper. maybe @pltrdy can elaborate.
closing in the meantime.
